[1단계 - 출석 구현] 랜디(김정우) 미션 제출합니다.

KJungW opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3906


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

아직 리펙토링을 하지못한 부분이 많아 객체지향 생활체조 요구사항을 많이 충족하지 못했다고 생각합니다.
else 예약어를 쓰지 않는다와 같은 기본적인 요구사항은 지켰지만 아직 많은 리펙토링이 필요하다고 생각합니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?
- TDD의 진행방식에 대해 궁금합니다!
: 이번 미션에서는 아주 간단하게 설계를 하고난 뒤, 클래스마다 실패테스트를 작성하고 개발하는 방식으로 TDD를 진행했습니다. 하지만 해당 방식에서 TDD의 이점이 전혀 느껴지지 않고 번거롭다는 생각이 많이 들었습니다. 특히 TDD를 통해 설계를 해나간다는 느낌이 아니라 무리하게 실패 테스트를 만들어낸다는 느낌이 많이 들었습니다.
이 문제를 해결하기 위해 다음 TDD를 할 때는 극단적으로 사전 설계 하나 없이 중심 요구사항을 하나씩 선택해서 TDD를 진행하는 방향을 생각해보았습니다. 이를 통해 큰도메인으로 개발을 시작해서 점차 작은 도메인들로 나눠가는 형태로 TDD 진행을 해볼까 생각 중인데 이번에 리뷰를 맡아주신 조앤의 의견이 궁금합니다! 그리고 조앤은 TDD를 어떤 방식으로 진행하시는지도 궁금합니다!

taek2222 committed:
docs(readme): 기능 구현 목록 초안 작성

taek2222 committed:
docs(readme): 기능 구현 목록 초안 수정

- 마크 다운 형식의 오해로 취소선 및 구분선 삭제했습니다.

taek2222 committed:
test(AttendanceWarningTypeTest): 테스트 추가

- 출결 상황에 따라 알맞은 경고를 반환하는 테스트를 추가 했다.

taek2222 committed:
feat(AttendanceWarningType): 타입 조회 기능 추가

- 출결 상황에 알맞게 타입을 찾아 반환한다.

- 단, 3회 지각은 1회 결석으로 처리한다.

taek2222 committed:
test(AttendanceWarningTypeTest): 테스트 리팩토링

- 출결 여러 상황에 알맞는 파라미터 테스트 형태로 변경했습니다.

taek2222 committed:
docs(readme): 구현 목록 체크

KJungW committed:
test(AttendanceStatusTypeTest): 테스트 추가

- 임계값에 해당하는 출결 상태를 반환한다

KJungW committed:
feat(AttendanceStatusType): 기능 추가

- AttendanceStatusType 생성 및 조회 구현

KJungW committed:
docs(readme): 구현 완료 기능 체크

taek2222 committed:
test(EducationTimeTest): 테스트 추가

- 월요일에 알맞은 지각 시간을 반환하는 테스트를 설계했다.

taek2222 committed:
feat(EducationTime): 지각 시간 계산 기능 추가

- 등교 시각과 시작 시각을 비교해 지각 시간을 반환한다.

taek2222 committed:
test(EducationTimeTest): 테스트 추가

- 요일 별로 여러 시간대의 파라미터 테스트를 추가했다.

taek2222 committed:
fix(EducationTime): 이전 시간 출석 버그 수정

- 절댓값으로 잘못된 차이를 계산해 출석한 경우의 차이값을 0으로 반환하게 했다.

KJungW committed:
test(CampusTimeTest): 테스트 추가

- 캠퍼스 운영시간이 아니라면 예외를 발생한다

KJungW committed:
feat(CampusTime): 캠퍼스 시간 구현

- 캠퍼스 운영 시간인지 검증하는 기능을 구현한다.

KJungW committed:
docs(readme): 구현 기능 목록 수정

- 구현된 기능 체크 표시
- 예외 상황 추가

KJungW committed:
test(CampusTimeTest): 테스트 리펙토링

- "캠퍼스 운영시간이 아니라면 예외를 발생한다."
  테스트 케이스 추가

taek2222 committed:
test(HolidayTest): 테스트 추가

- 등교일이 아닌 경우를 확인할 수 있는 테스트를 설계했다.

taek2222 committed:
feat(Holiday): 공휴일 검증 기능 추가

taek2222 committed:
test(HolidayTest): 테스트 추가

- 등교일 아닌 경우의 파라미터 테스트를 추가했다.

- 휴일 추가가 중복될 경우 예외 발생을 추가했다.

taek2222 committed:
feat(Holiday): 공휴일 추가하는 기능 추가

- 확장성을 고려해 공휴일을 추가하는 메서드를 추가했다.

- 공휴일도 휴일로 판정될 수 있도록 리팩토링 했다.

KJungW committed:
test(AttendanceTest): 테스트 추가

- 기본 날짜와 시간으로 출석을 생성할 수 있다

KJungW committed:
feat(Attendance): 디폴트 출석 기록 생성

- 입력된 날짜에 해당하는 디폴트 출석 기록을 생성한다.

KJungW committed:
test(AttendanceTest): 테스트 추가

- 기본 날짜와 시간으로 출석을 생성할 수 있다

KJungW committed:
feat(Attendance): 출석 시간 업데이트 구현

- 입력된 시간으로 출석 시간을 업데이트한다

KJungW committed:
test(AttendanceTest): 테스트 추가

- 출석 시간을 수정할 수 있다

taek2222 committed:
test(AttendancesTest): 테스트 추가

- 크루가 추가될때 기본 출석이 생성되는 테스트를 구현했다.

taek2222 committed:
feat(Attendances): 크루를 추가하는 기능

- 크루를 기본적으로 생성하는 로직을 설계했다.

- 크루 이름으로 찾을 수 있는 일정을 반환하는 메서드도 설계했다.

KJungW committed:
refactor(Attendance): 날짜 생성기 적용

- 테스트하기 어려운 날짜 생성 부분 분리

KJungW committed:
test(AttendancesTest): 테스트 수정

- 모킹 날짜 생성기를 주입해서 테스트하도록 수정한다.

taek2222 committed:
test(AttendanceTest): 테스트 추가

- 동일 날짜 여부에 따라 반환하는 값을 추가한다.

taek2222 committed:
feat(Attendance): 날짜 동일 여부 메서드 추가

taek2222 committed:
feat(Attendances): 출석 등록 기능 구현

- 애플리케이션 셋팅을 위해 설정을 주입했습니다.

- 입출력 부분을 정의해 비지니스 로직에 추가했습니다.

KJungW committed:
refactor(AttendanceManager): 기존 클래스명 수정

- Attendances를 AttendanceManager로 수정

KJungW committed:
test(AttendancesTest): 테스트 추가

- 출석을 추가할 수 있다
- 해당 날짜의 출석을 찾을 수 있다

KJungW committed:
test(HolidayTest): 테스트 실패 수정

- 공휴일 세팅값을 추가해서 테스트가 성공하도록 수정한다

KJungW committed:
refactor(Attendances): Attendances 적용

- 코드의 단순화를 위해 출석 리스트를 묶는 Attendances를 적용한다

KJungW committed:
feat(AttendanceService): 출석 수정 구현

- 기존의 출석 기록을 수정할 수 있다

taek2222 committed:
feat(attendances.csv): 출석 초기 데이터 생성

taek2222 committed:
feat(FileUtil): 파일 입력 유틸 추가

taek2222 committed:
feat(AttendanceService): 초기 크루 출석 설정 및 적용

KJungW committed:
feat(AttendanceService): 회원의 출석기록 출력

- 회원의 출석기록을 출력한다

taek2222 committed:
feat(AttendanceWarningType): 출석 이름 타입 추가

taek2222 committed:
feat(AttendanceService): 크루 출석 목록에 데이터 추가

KJungW committed:
feat(AttendanceService): 제적 위험자 조회 구현

- 제적 위험성이 있는 학생들을 조회한다

taek2222 committed:
docs(*): TodoList 주석 작성

taek2222 committed:
refactor(Attendances): 출석 저장 형태 Map -> 필드 전환

- 기능 변동으로 4번 기능의 오류로 주석처리 했다.

taek2222 committed:
refactor(Attendances): 오늘 날짜 까지만 기록 조회로 변경

- 출석 기록을 오늘 날짜까지만 출력할 수 있도록 변경했다.

taek2222 committed:
docs(readme): 기능 구현 완료 체크

KJungW committed:
feat(AttendanceManager): 제적 위험자 정렬

- 제적 위험자을 요구사항에 주어진 기준에 따라 정렬한다

KJungW committed:
docs(readme): 완료된 요구사항 체크

taek2222 committed:
refactor(AttendanceController): 재실행 기능 추가

- 필드 관리를 위해 메뉴 커맨드를 Enum으로 분리했다.

- 일치하지 않은 커맨드는 에러가 발생하도록 기능을 추가했다.

KJungW committed:
feat(AttendanceController): 메뉴 반복 입력 추가 적용

- 메뉴의 모든 기능에 대해 메뉴 반복 입력 추가 적용

KJungW committed:
feat(AttendanceManager): 등록되지 않은 닉네임 처리

- 등록되지 않은 닉네임이 임력될 경우 예외 처리

KJungW committed:
refactor(AttendanceManager): 등록된 크루 확인 수정

- 등록된 크루가 아닌 경우 true를 리턴한다.

KJungW committed:
feat(Attendance): 이미 출석된 경우 처리

- 이미 출석되었는데 출석을 시도하는 경우 예외처리

KJungW committed:
fix(AttendanceController): 메뉴 반복 버그 수정

- 무한 반복 버그 수정

taek2222 committed:
feat(InputView): 최소한 검증 로직 추가

- 검증 로직 추가로, 반환 값을 변환 형태로 반환하도록 변경했다.

taek2222 committed:
feat(AttendanceInitService): 초기 설정 서비스 분리

- DateTime 파싱 유틸 클래스를 정의해 에러까지 처리하도록 했다.

KJungW committed:
refactor(AttendanceService): 날짜와 시간 파싱에 에외처리 추가

- 날짜와 시간 파싱에 예외 처리를 추가한다.

taek2222 committed:
refactor(AttendanceService): 오늘까지 기록 반환 로직 도메인으로 변경

KJungW committed:
refactor(AttendanceService): 메서드 분리

- 코드의 단순화를 위해 메서드 분리를 수행한다

KJungW committed:
docs(readme): 구현 완료한 기능 체크

KJungW committed:
refactor(AttendanceManager): 코드 단축

- 스트림을 활용해 기존의 코드를 단축한다

taek2222 committed:
fix(AttendanceManager): 마지막 날까지 생성 안되는 버그 수정

taek2222 committed:
refactor(Attendances): 출석 줄바꿈 개선

KJungW committed:
refactor(AttendanceService): 입력 버그 수정

- 등교 시간을 입력받을때 잘못된 안내문구가 나오는 것을 수정

KJungW committed:
fix(AttendanceService): 닉네임 입력 버그 수정

- 학생의 닉네임을 입력받을때 다른 안내문구가 뜨는 버그 수정

KJungW committed:
refactor(OutputView): 출력문 줄바꿈 추가

- 필요한 경우 출력문에 출바꿈 추가

seovalue commented on README.md:
```diff
@@ -1,3 +1,273 @@
 # java-attendance
 
-출석 미션 저장소
+## :sparkling_heart: 프로젝트 소개
```
> TDD의 진행방식에 대해 궁금합니다!

저도 랜디(ssg 팬이신가요?) 의견에 완전 공감해요. 그래서 TDD가 어렵다고 느껴지고요ㅎㅎ
제가 TDD를 해야할 일이 있다면 말씀해주신대로 극단적으로 테스트에서 시작하고 도메인을 채워나가는 순서로 진행하는 것 같아요. 특히 크리티컬한 도메인 로직 위주로요.

KJungW replied:
생각했던 방향이 올바른 방향인 것 같아 다행인 것 같아요! 나중에 2단계할때는 설계없이 주요 요구사항을 하나씩 골라서 TDD 개발을 진행해보겠습니다! (ssg는 잘모르겠습니다..!)

seovalue commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+## 출석 파일 입력
+
+### 출석 파일 입력
```
👍 요구사항 잘 작성해주셨네요

KJungW replied:
감사합니다! 추가적으로 대제목과 소제목이 똑같이 겹치는 부분은 가독성을 위해 대제목 앞부분에 이모지를 추가해 넣었습니다!

seovalue commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Set;
+import java.util.stream.IntStream;
+
+public class AttendanceManager {
```
AttendanceManager가 정확히 어떤 역할을 하는거예요?

KJungW replied:
출석을 관리하는 클래스라는 생각에 AttendanceManager라는 이름을 작성했습니다. 하지만, 조앤이 짚어주시고 다시 곰곰히 생각해보니 AttendanceManager라는 이름이 해당 클래스 하는 일을 명확하게 표현하지 않고있다는 생각이 들었습니다. 그리고 다시 그 이유를 생각해보니 AttendanceManager가 너무 많은 역할을 맡고 있다는 것을 깨닫게 되었습니다. 그래서 AttendanceManager를 역할 별로 분리해서 관리하는 방향으로 재설계를 해보았습니다!

seovalue commented on src/main/java/attendance/utility/DateGenerator.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.time.LocalDate;
+
+public interface DateGenerator { // todo : 공백 제거
```
DateGenerator는 내부에서 LocalDate.now()만 수행하는 것 같은데 인터페이스화가 필요한 이유가 궁금해요!

KJungW replied:
당일에 대한 LocalDate를 가져오는 부분을 추상화해서 이후 테스트를 할때 해당 부분을 모킹하려고 했습니다! 하지만 설계를 수정하다보니 크게 필요하지 않다는 생각이 들어 제거했습니다!

seovalue commented on src/main/java/attendance/service/AttendanceService.java:
```diff
@@ -0,0 +1,87 @@
+        this.attendanceManager = attendanceManager;
+    }
+
+    public AttendanceResponse processAttendance(LocalDate today) {
```
전체적으로 네이밍에 `process`를 붙여주셨던데, 각 메서드별로 좀 더 역할이 잘 드러나는 네이밍을 하면 어떨까요?
아래 processUpdateAttendance도 사실 updateAttendance가 되어도 무방한 것 같거든요. 꼭 프로세스를 포함시키지 않아도 될 것 같은데, 어떻게 생각해요??

KJungW replied:
컨밴션 측면에서 서비스의 메서드라는 것을 표현하기 위해 process라는 이름을 접두사로 붙여야 한다고 생각했는데 너무 관성적인 생각이었던 것 같습니다. 컨밴션도 중요하지만 조앤이 말씀해주신 대로 컨밴션보다는 명확한 의미전달이 더욱 중요할 것 같다는 생각이 듭니다. 다른 부분에서도 이러한 문제가 있는지 확인해보고 수정해보겠습니다!

seovalue commented on src/main/java/attendance/service/AttendanceService.java:
```diff
@@ -0,0 +1,87 @@
+
+public class AttendanceService { // todo : 최대한 메서드 분리 작업
+
+    private final InputView inputView;
```
Service 내부에 InputView가 있는거에 대해서 어떻게 생각해요?

KJungW replied:
서비스는 컨트롤러의 연장선이라는 아이디어를 가지고 이번 미션에서 처음으로 중심 로직 서비스에 View를 주입해 사용해보았습니다. 시도는 좋았지만 아쉽게도 미션이 진행될수록 입출력 파트에 의해 리펙토링도 어려워지고, 코드도 깔끔하지 않다는 생각이 들었습니다. 이 부분을 리펙토링 대상으로 생각하고 있었는데 조앤이 짚어주신 덕분에 좀더 확신을 가지고 리펙토링 할 수 있을 것 같습니다!

seovalue commented on src/main/java/attendance/service/AttendanceService.java:
```diff
@@ -0,0 +1,87 @@
+        return attendanceManager.searchWarnedCrews(today);
+    }
+
+    private Attendances findAttendance(boolean isForUpdated) {
+        String nickname = inputView.readNickname(isForUpdated);
+        return attendanceManager.findCrewAttendance(nickname);
```
isForUpdate가 어떻게 동작하는지 아려면 inputView까지 들어가야하는데, 분리할 수 없을까요?

KJungW replied:
이 부분도 서비스에 입출력 파트가 주입되며 발생한 문제인 것 같습니다! 입출력 파트에 의해 서비스의 메서드가 길어지고, 길어진 메서드를 줄이기 위해 조금 무리한 방식을 선택했었는데 Service와 View를 분리하면서 이 부분도 함께 해결해보겠습니다!

seovalue commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,107 @@
+        printBlankLine();
+    }
+
+    public void printWarnedStudents(WarnedStudentResponses response) {
```
어떤건 파라미터로 response를 어떤건 result를 받는데 통일해보아요~!

KJungW replied:
확실히 일관성에 대한 부분을 놓쳤었던 것 같습니다. 다른 것들도 확인해보고 함께 수정해보겠습니다!

seovalue reviewed (changes requested):
안녕하세요 랜디~
몇가지 코멘트 남겼는데 확인 후 리뷰 요청 주세요!
오늘도 화이팅~!

KJungW committed:
feat(crew): 크루 저장소 도메인 추가

크루 저장소 도메인
- 크루 추가 기능 구현
- 등록된 크루 확인 기능 구현

테스트 내용
- 크루를 추가할 수 있다
- 이름을 통해 크루가 등록된 크루인지 확인할 수 있다

KJungW committed:
feat(crew): 크루 도메인 추가

크루 도메인
- 크루 이름을 포장하는 원시값 포장 객체
- 이름을 기반으로 equals 정의

테스트 내용
- 이름이 같으면 동일한 크루로 취급한다
- 크루 이름은 공백을 허용하지 않는다

KJungW committed:
refactor(crew): 크루 도메인 적용

작업 내용
- 크루 도메인을 크루 저장소에 적용

KJungW committed:
feat(holiday): 휴일 판별기 도메인 생성

휴일 판별기 도메인
- 휴일 추가 기능
- 휴일 판별 기능

테스트 내용
- 공휴일을 추가할 수 있다
- 주말인 경우 휴일로 간주한다
- 주말이 아닌 공휴일인 경우 휴일로 간주한다
- 주말도 아니고 공휴일도 아닌 경우 휴일로 간주하지 않는다

KJungW committed:
feat(schedule): 캠퍼스 스케줄 도메인 추가

캠퍼스 스케줄 도메인
- 캠퍼스 운영 시간인지 검증하는 기능
- 출석인지 확인하는 기능

테스트 내용
- 캠퍼스 운영시간이 아닌 경우 예외를 발생시킨다.
- 월요일인 경우의 출석여부를 구한다
- 월요일이 아닌 경우의 출석여부를 구한다

KJungW committed:
feat(record): 출석 기록 저장소 도메인 추가

출석 기록 저장소 도메인
- 출석 기록 저장
- 출석 기록 업데이트
- 출석 기록 조회

테스트 내용
- 크루에 대한 출석 기록을 추가한다
- 결석일 경우에는 출석 기록 추가가 불가능하다
- 출석 기록을 수정할 수 있다
- 출석 기록을 결석으로 수정할 경우 기존의 출석 기록을 제거한다

KJungW committed:
refactor(service): 기존의 출석 시비스 제거

제거 이유
- 입출력 뷰와 혼합되어 있어 다시 구현 예정

KJungW committed:
docs(readme): 요구사항 수정

수정 내용
- 현재 요구사항과 관련 없는 내용 제거

KJungW committed:
feat(system): 출석 시스템 도메인 구현

출석 시스템 도메인
- 출석 저장 구현
- 출석 수정 구현
- 크루별 출석 조회 구현

테스트 내용
- 출석 저장 - 닉네임과 등교 시간을 입력받아 출석 기록 저장
- 출석 저장 - 결석 기록은 저장되지 않는다
- 출석 저장 - 캠퍼스 운영시간이 아닌 경우 예외 발생
- 출석 저장 - 이미 출석을 완료한 경우 예외 발생
- 출석 저장 - 휴일의 경우 예외 발생
- 출석 저장 - 등록되지 않은 닉네임의 경우 예외 발생
- 출석 수정 - 출석 기록을 수정한다
- 출석 수정 - 결석 기록으로 수정할 시 출석 기록이 제거된다
- 출석 수정 - 캠퍼스 운영시간이 아닌 경우 예외 발생
- 출석 수정 - 휴일의 출석 기록 수정시 예외 발생
- 출석 수정 - 등록되지 않은 닉네임의 경우 예외 발생
- 출석_조회_크루별_출석을_조회할_수_있다
- 출석 조회 - 등록되지 않은 닉네임의 경우 예외 발생

KJungW committed:
feat(system): 제적 위험자 조회 구현

테스트 내용
- 제적 위험자 조회 - 제적 위험자를 조회할 수 있다.

KJungW committed:
refactor(application): 애플리케이션 정리

작업 내용
- 사용하지 않는 클래스 제거
- 클래스명과 필드명을 적절히 수정

KJungW committed:
refactor(initailization): 변경된 구조에 맞게 초기화 구현

작업 내용
- 기존의 초기화를 수정해서 변경도니 구제에 맞게 초기화 구현

KJungW committed:
refactor(record): 월별 출석 기록 조회시 연도 고려

작업 내용
- 모든 연도의 월별 출석 기록이 조회되지 않도록
  월별 출석 기록을 조회할 때 연도를 고려하도록 수정

KJungW committed:
feat(system): 제적 위험 정보 단건 조회

작업 내용
- 하나의 크루에 대한 제적 위험 정보 출력 구현

테스트 내용
- 출석 상태 조회 - 해당 크루의 제적 위험 정보를 조회한다
- 출석 상태 조회 - 등록되지 않은 닉네임의 경우 예외 발생

KJungW committed:
refactor(risk): 제적 위험성의 필드 수정

작업 내용
- 출석 횟수 필드 추가
- 필드명을 보다 적절하게 수정

KJungW committed:
feat(input_output): 입출력 구현

작업 내용
- 요구사항에 맞게 입출력 구현

KJungW committed:
refactor(input_output): 요구사항에 맞게 입출력 형식 수정

작업 내용
- 요구사항에 맞지 않게 출력되는 내용들을 수정

KJungW committed:
refactor(exception): 예외메세지 상수화

작업 내용
- 리터럴이었던 예외메세지 상수화

KJungW committed:
refacotr(system): 출석 시스템 도메인의 역할 분리

작업 내용
- 기존의 출석 시스템에서 하던 일을
  다른 도메인에게 적절히 분배

KJungW committed:
style(application): 공백 정리

KJungW committed:
docs(readme): 요구사항 정리

KJungW committed:
fix(system): 출석 조회는 당일까지만 조회

작없 내용
- 아직 지나지 않은 날이 결석 처리된는 버그 수정

KJungW commented:
## 1단계 미션 첫번째 피드백을 받은 이후 추가로 수정한 점

### 대부분의 DTO 제거, 대신 도메인에 getter를 만들어 출력뷰에서 활용하도록 수정
- 처음에는 뷰와 도메인이 서로 완전히 의존하지 않도록 도메인과 뷰를 완전히 분리해보자는 생각에 DTO를 적용시켜보았습니다. 하지만 오히려 DTO로 인해 코드가 복잡해지며 득보다는 실이 많다는 생각이 들었습니다. 그래서 도메인은 뷰를 의존하지 않지만, 뷰는 도메인에 대해 의존 가능하도록 다시 설계를 해보았습니다. 그리고 DTO는 도메인에서 출력뷰로 여러 데이터들을 묶어서 전달해야 하는 경우에만 사용하려고 합니다!

### 설계를 수정하며 불필요해진 서비스 계층 제거
- 서비스와 뷰를 분리하다보니 서비스 계층의 존재 이점이 없다는 생각이 들었습니다. 그래서 과감하게 서비스 계층을 제거하고 관리 시스템 전체를 총괄하는 AttendanceSystem 도메인 클래스를 만들어 도메인 중심 설계를 진행해보았습니다!

### 도메인에 대한 테스트 추가
- 미션을 진행하며 놓쳤거나 부족하다고 생각되는 테스트들을 추가해보았습니다.

seovalue commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,99 @@
+    private final AttendanceSystemInitializer initializer;
+
+    public AttendanceController(AppConfig appConfig) {
+        this.inputView = appConfig.getInputView();
+        this.outputView = appConfig.getOutputView();
```
```java
this.inputView = new InputView();
this.outputView = new OutputView();
```
appConfig에서 꼭 전체적으로 생성해주는 이유는 뭐예요?

KJungW replied:
가장 큰 이유는 전체적인 의존성 주입의 흐름을 한 곳에 모아서 관리할 수 있다는 점인 것 같습니다! 상위 클래스에서 객체를 생성해 하위 클래스로 직접 주입하는 방식은 의존성 주입의 흐름이 여러 곳에 퍼져있어 한눈에 살펴보기 힘들다는 단점이 있었습니다. 컨트롤러에서 전체적인 의존성 주입을 했을 때는 컨트롤러 코드가 깔끔하지 않다는 생각이 들었습니다. 그래서 시스템의 전반에 걸친 전체 의존성 주입을 전담하는 Config 객체를 만들어 사용하고 있습니다!

seovalue commented on README.md:
```diff
@@ -1,3 +1,259 @@
+
+### 출석 파일 입력
+
+- [x] 출석 파일 입력
```
랜디! 구현해주신 부분 보면 전체적으로 객체가 많이 나뉘어져있는데, 각 객체 별로 역할을 정리해주시면 좋을 것 같아요.
- AttendanceRecord:
- CrewStorage:
- ...

KJungW replied:
확실히 짚어주신 부분을 생각해보니 내부 클래스 파일들이 많이 번잡하다는 생각이 들었습니다! 주요 키워드를 통해 패키지로 그룹화 시켜보겠습니다!

seovalue commented on src/main/java/attendance/utility/FileUtility.java:
```diff
@@ -0,0 +1,32 @@
+import java.io.FileReader;
+import java.util.List;
+
+public class FileUtility {
```
File을 다루는 용도의 util을 나눈 것 저는 좋아보이네요!

KJungW replied:
페어의 의견이었는데 확실히 좋은 의견이었다고 생각합니다! 설계에 따라 다르겠지만 다음부터도 파일을 다루는 유틸리티 클래스를 따로 만들게 될 것 같습니다.

seovalue commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,125 +1,99 @@
-            }
-        }
+    private void processMenuFunction(AttendanceMenu menu) {
+        MenuTemplate.run(this::saveAttendanceRecord, menu, outputView);
+        MenuTemplate.run(this::updateAttendanceRecord, menu, outputView);
+        MenuTemplate.run(this::searchAttendanceRecordsByCrew, menu, outputView);
+        MenuTemplate.run(this::searchRiskStatistics, menu, outputView);
```
method line 수 때문에 이렇게 구현해주신 건가요?
뭔가 Menu는 단순 입력을 받아 알맞은 메서드를 실행시키는 역할인데 MenuTemplate까지 필요한 이유가 궁금해요!

KJungW replied:
프로그램 요구사항에 메서드 길이 10줄 제한이 있어 코드를 분리하기는 했는데 확실히 가독성을 떨어뜨리는 불필요한 작업이었던 것 같습니다! 템플릿을 제거해서 가독성을 높여보겠습니다!

seovalue commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,125 +1,99 @@
+        LocalDate nowDate = LocalDate.now();
+        if (menu == AttendanceMenu.UPDATE) {
+            String nickname = inputView.readNicknameForUpdate();
+            LocalDate date = nowDate.withDayOfMonth(inputView.readDateForUpdate());
```
오홍 withDayOfMonth라는 건 저도 첨 알았네요 👍

seovalue commented on src/main/java/attendance/domain/AttendanceRecord.java:
```diff
@@ -0,0 +1,73 @@
+        return arrivalDateTime.getYear() == year && arrivalDateTime.getMonth() == month;
+    }
+
+    public boolean isInPeriod(LocalDate startDate, LocalDate endDate) {
+        LocalDate date = arrivalDateTime.toLocalDate();
+        boolean isAfterOrEqual = date.isAfter(startDate) || date.equals(startDate);
+        boolean isBeforeOrEqual = date.isBefore(endDate) || date.equals(endDate);
+        return isAfterOrEqual && isBeforeOrEqual;
+    }
```
```suggestion
public boolean isInPeriod(LocalDate startDate, LocalDate endDate) {
        return !arrivalDateTime.toLocalDate().isBefore(startDate) &&
                !arrivalDateTime.toLocalDate().isAfter(endDate);
    }

```
이런 식으로도 줄여볼 수 있을 것 같아요!
지금은 조건이 너무 많은 것 같아서..

KJungW replied:
훨씬 깔끔한 것 같습니다! 바로 리펙토링 해보겠습니다!

seovalue commented on src/main/java/attendance/domain/AttendanceRecordStorage.java:
```diff
@@ -0,0 +1,75 @@
+
+    public void add(AttendanceRecord record) {
+        validateNotSaved(record.getNickname(), record.getDate());
+        if (!record.isExpulsion()) {
+            List<AttendanceRecord> records = findRecordsByNickname(record.getNickname());
+            records.add(record);
+            attendanceRecords.put(record.getNickname(), records);
+        }
```
if (record.isExpulsion()) return
이렇게 early-return을 사용해보는 건 어때요?

KJungW replied:
지정된 컨밴션에 의해 바로 return을 진행하는 if문이라도 최소 세줄로 작성되어서 메서드 길이를 조금이라도 줄여보고자 early-return을 사용하지 않았습니다! 지금 다시 생각해보면 컨벤션은 코드의 가독성을 높이고자 존재하는데 컨밴션을 지키는 것에만 너무 집중한 나머지 오히려 코드의 가독성이 떨어트린 것 같습니다. 이제부터라도 컨밴션보다 가독성에 우선순위를 두고 early-return을 사용해보겠습니다!

seovalue commented on src/main/java/attendance/domain/AttendanceRecordStorage.java:
```diff
@@ -0,0 +1,75 @@
+
+public class AttendanceRecordStorage {
+
+    private final Map<String, List<AttendanceRecord>> attendanceRecords = new HashMap<>();
```
value인 List<Attendance>도 일급컬렉션화해보는 건 어때요?
Map<CrewName, Attendances>

KJungW replied:
객체지향적인 관점에서 좋은 설계인 것 같습니다! 바로 리펙토링 해보겠습니다!

seovalue commented on src/main/java/attendance/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class AttendanceSystem {
```
AttendanceSystem과 Controller는 무슨 차이예요?

KJungW replied:
AttendanceSystem은 전체 도메인 로직들을 감싸서 최종적인 인터페이스 역할을 하는 도메인으로 설계하였습니다. 그래서 외부에서 도메인 로직을 활용하고자 할때는 AttendanceSystem이 제공하는 인터페이스를 활용하도록 구현하였습니다.  반면에 컨트롤러는 전체 도메인 로직을 감싼 AttendanceSystem와 입출력을 담당하는 View 사이를 연결하고 실행하면서 전체적으로 애플리케이션을 실행하는 역할로 설계하였습니다!

seovalue replied:
이해했어요! 동의합니다

seovalue commented on src/main/java/attendance/domain/HolidayChecker.java:
```diff
@@ -0,0 +1,45 @@
+
+public class HolidayChecker {
+
+    private final List<LocalDate> publicHolidays = new ArrayList<>();
```
holiday 관리를 List로 하신 이유가 궁금해요!

KJungW replied:
확실히 고민없이 정했던 관성적인 설계였던 것 같습니다. 휴일을 저장하는 순서를 따로 고려할 필요가 없고, 휴일의 중복저장을 허용하지 않는 조건에서 List보다 Set으로 관리하는 것이 적절하다는 생각이 듭니다! 바로 리펙토링 해보겠습니다!

seovalue commented on src/main/java/attendance/domain/RiskStatistic.java:
```diff
@@ -0,0 +1,38 @@
+package attendance.domain;
+
+public final class RiskStatistic {
```
이건 View에 전달하기 위한 dto 같은 용도일까요?

KJungW replied:
저도 DTO 성격이 강하다는 생각이 들어 DTO로 설계를 했으나, 생성자에서 결석횟수와 지각횟수를 받아 제적 위험도 타입을 계산하는 비즈니스 로직이 포함되어 있어 최종적으로는 도메인으로 결정하게 되었습니다. 이 부분에 대해서 저도 애매하다는 생각이 많이 들었는데 이런 경우에 조앤은 보통 어떻게 처리하시는지 궁금합니다!

seovalue replied:
말씀주신 내용이 RiskType의 로직이라 이건 dto가 되어도 무방할 것 같은 느낌이 드네요!!

seovalue commented on src/main/java/attendance/view/InputView.java:
```diff
@@ -2,42 +2,42 @@
 import java.util.Scanner;
 
-public class InputView { // todo : 상수 분리, 최소한의 검증 후 반환
+public class InputView {
```
전에 제출해주셨을 때 보다 View와 Domain 간의 강결합이 확실히 제거됐네요! 👍

KJungW replied:
확실히 도메인 로직도 깔끔해지고 관리도 편해진 것 같습니다!

seovalue reviewed (changes requested):
안녕하세요 랜디! 
이전보다 깔끔해졌네요 👍 
몇가지 질문 남겨뒀는데 확인 후 답변/리뷰 요청 남겨주세요!

KJungW committed:
refacotr(application): 패키지 분리

작업 내용
- 가독성을 위해 패키지를 분리해서 관리

KJungW committed:
refactor(menu): 템플릿 제거

작업 내용
- 가독성을 위해 템플릿 제거

KJungW committed:
refactor(holiday): 주말이 아닌 공휴일 저장 구조 변경

작업 내용
- List가 아닌 Set을 사용함으로써 중복 저장 위험을
  원천 봉쇄

KJungW committed:
refactor(record): 크루별 출석 기록 리스트를 추상화

잡업 내용
- 크루별 출석 기록을 별도의 클래스로 추상화하여
  객체지향성을 높임

KJungW committed:
test(application): 패키지 분리

작업 내용
- 프로덕션 코드에 맞게 테스트 코드도 패키지 분리

KJungW committed:
test(fixer): 출석 기록 생성기 추가

작업 내용
- 테스트를 위한 출석 기록을 생성하는 생성기 구현

KJungW committed:
refactor(record): 기한 체크 조건 간소화

작업 내용
- 기한 체크 조건을 간소해서 가독성 증대

seovalue commented on src/main/java/attendance/domain/record/AttendanceRecord.java:
```diff
@@ -31,10 +31,8 @@ public boolean isInMonth(int year, Month month) {
     }
 
```
변경점이 아니여서 이상한 줄에 코멘트 남기게 됐는데ㅋㅋ
저는 Crew(name)가 이미 있는데도 여기서 String nickname을 가져가는게 관리포인트를 늘리는 결정같이 느껴져요.
랜디는 이부분에 대해서 어떻게 생각하시는 지 궁금해요!

seovalue replied:
AttendanceRecordStorage 여기서의 key 값도 여전히 string이고요!

KJungW replied:
우선 크루 데이터의 일관성 유지가 가장 큰 이유인 것 같습니다! CrewStorage에서 등록된 크루들을 전담으로 저장하고 관리하는 상태에서 AttendanceRecordStorage에서도 내부에 Crew를 Map의 key로써 저장한다면 두 곳에서 같은 Crew를 저장하고 있기 때문에 일관성 문제가 발생할 수 있다고 생각했습니다. 현재 시스템 기준에서는 크루가 내부에 name필드만 존재하고 불변객체처럼 사용되고 있어 두 곳에서 저장해도 큰 문제없지만,  Crew 도메인은 이후 설계 변경으로 가변 필드(ex. 현재 출석 상황에 따른 제적위험도)가 생성될 수 있다고 생각했고 이로인해 일관성 문제가 발생할 수 있을 것 같다고 판단했습니다. 그래서 Crew 도메인을 불변객체아닌 일반객체로 설계하면서 AttendanceRecordStorage에서는 nickname으로 크루를 표시하게 되었습니다.  결과적으로 저는 Crew 도메인의 확장성과 무결성을 위해 관리포인트를 늘릴 가치가 있는 것 같다고 조심스럽게 의견을 내보겠습니다!

seovalue reviewed (approved):
안녕하세요 랜디!! 
추가적으로 코멘트 두개 남겨뒀는데 랜디 생각이 궁금해요ㅎㅎ
step2 진행하면서 반영해주시면 좋을 것 같고, step2는 TDD로 모든 요구사항을 구현해보시고 이를 커밋로그로 확인할 수 있도록 해주시면 좋을 것 같아요~!
